20240917-BR-resiliency-error-code-retries.md

```diff
@@ -0,0 +1,149 @@
+Currently, all errors are retried when the policy is applied.
+Some status codes are not retryable, and subsequent calls will result in the same error. Avoiding these retry calls will reduce the overall number of requests, traffic, and errors.
+
+## Related Items
+
+https://github.com/dapr/dapr/issues/6683
+https://github.com/dapr/dapr/issues/6428
+https://github.com/dapr/dapr/issues/7697
+
+PR:
+https://github.com/dapr/dapr/pull/7132
+
+Docs:
+https://github.com/dapr/docs/issues/4254
+https://github.com/dapr/docs/issues/3859
+
+## Expectations and alternatives
+
+* What is in scope for this proposal?
+  - HTTP and gRPC Service Invocation, direct and proxied
+  - Bindings
+  - Pub/Sub
+
+## Implementation Details
+
+### Design
+
+Add a new object field to the `retries` policy Spec to allow the user to specify the status codes that should be retried.
+Separate fields for HTTP and gRPC. The new fields should be optional and will default to the existing behavior, which is to retry on all errors.
+
+### Example 1:
+In this example, the retry policy will retry **_only_** on HTTP 500 and HTTP status code range 502-504 (inclusive) and gRPC status code range 2-4 (inclusive).
+The rest of the status codes will not be retried.
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Resiliency
+metadata:
+  name: myresiliency
+scopes:
+  - app1
+spec:
+  policies:
+    retries:
+      pubsubRetry:
+        policy: constant
+        duration: 5s
+        maxRetries: 10
+        matching:
+          httpStatusCodes: "500,502-504"
+          gRPCStatusCodes: "2-4"
+```
+
+### Example 2:
+In this example, the retry policy will retry **_only_** on gRPC status code range 1-15 (inclusive).
+However, this policy will not apply to the HTTP status codes, and they will be retried according to the default behavior, which is to retry on all errors.
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Resiliency
+metadata:
+  name: myresiliency
+scopes:
+  - app1
+spec:
+  policies:
+    retries:
+      pubsubRetry:
+        policy: constant
+        duration: 5s
+        maxRetries: 10
+        matching:
+          gRPCStatusCodes: "1-15"
+```
+
+### Acceptable Values
+The acceptable values are the same as the ones defined in the [HTTP Status Codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) and [gRPC Status Codes](https://grpc.io/docs/guides/status-codes/) documentation.
+
+- HTTP: from 100 to 599
+- gRPC: from 1 to 16
+
+### Setting Format
+Both the `httpStatusCodes` and `gRPCStatusCodes` fields are of type string and optional and can be set to a comma-separated list of status codes and/or ranges of status codes.
+The range must be in the format `<start>-<end>` (inclusive). Having more than one dash in the range is not allowed.
+
+### CRD Validation
+
+Both field values should be validated using Common Expression Language [CEL](https://kubernetes.io/docs/reference/using-api/cel/)
```

Comment:
Nit: please can we also add reference to the kube builder docs? https://book.kubebuilder.io/reference/markers/crd-validation